gum/backend-libdwarf/gumsymbolutil-libdwarf.c

```diff
@@ -546,10 +622,10 @@ gum_collect_module_functions (const GumModuleDetails * details,
     return TRUE;
 
   gum_elf_module_enumerate_dynamic_symbols (entry->module,
-      gum_collect_symbol_if_function, NULL);
+      gum_collect_symbol_if_function, (gpointer) FALSE);
 
   gum_elf_module_enumerate_symbols (entry->module,
-      gum_collect_symbol_if_function, NULL);
+      gum_collect_symbol_if_function, (gpointer) TRUE);
```

Comment:
Can't we do both in the callback instead of walking the symbols twice?

Reply:
I'm not sure I folllow? The callback *gum_collect_symbol_if_function* now populates both data structures. We use it twice because there are two symbol tables to walk the one in .symtab and the other in .dynamic. We don't need to collect symbols from .dynamic into our reverse lookup as dladdr will work just fine on those. Hence we have the TRUE/FALSE parameter to avoid collecting unnecessary data.

Have I missed something?

Reply:
Oh! I'm so sorry, I misread the code. That means it should have been two separate callbacks like it was, right? (Unless we want to populate `gum_function_addresses` using both?)

Reply:
So previously, we walked both .dynamic and .symtab with *gum_collect_symbol_if_function* and then walked .symtab with the new callback. So we walked the .symtab twice.

Now we have a single combined callback with a parameter to indicate whether we want to collect the data into the second structure. So we only walk each of .dynamic and .symtab the once.

So we have the slight added complexity of an additional parameter, and a callback that does two different jobs. But in exchange we should have the performance increase of only processing the tables the once.

I think maybe it's better this new way?

Reply:
Ahh gotcha. Makes a lot of sense. Realize I was a bit knackered when doing the earlier review, so created quite a bit of confusion here. Sorry about that.

Reply:
No worries. We got there in the end.